### PR TITLE
Feature/2888 discussion settings

### PR DESCRIPTION
**Pull request by tonyr59h**
All new settings complete and some old ones re-visited for self-hosted sites. In order to test most of the self-hosted settings you'll have to update one of the files on your WordPress box. PM me for details.

cc @kwonye

### CODE REVIEW

File: WordPress/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -457,20 +480,62 @@
...
+    <string name="site_whitelist_hint">Comment author must have a previously approved comment</string>
+    <string name="site_multiple_links_hint">Ignores link limit from known users</string>
+    <string name="site_moderation_hold_hint">Comments that match a filter are put in the moderation queue</string>
+    <string name="site_blacklist_hint">Comments that match a filter are marked as spam</string>
     <string name="site_settings_related_posts_summary">Show related content after posts</string>
     <string name="site_settings_disconnected_toast">Disconnected, editing disabled.</string>
     <string name="site_settings_save_success">Site settings updated</string>
+    <string name="site_settings_learn_more_desc">In the Default article settings, there are three options. These settings are defaults for new posts or pages, which can always be changed individually on each article.</string>
+    <string name="site_settings_learn_more_bullets">The Allow Comment option allows you to enable or disable comments by default.</string>
+    <string name="site_settings_learn_more_caption">You can override thee settings for individual posts.</string>
```
Comment (kwonye) [inline WordPress/src/main/res/values/strings.xml]: typo "You can override thee settings"

**Comment by kwonye** [comment]
I noticed that there is no consistency to capitalization. There's "Allow Comments" and below there is "Links in comments".

Can you do an audit of all the newly committed text in `strings.xml` for this please?

**Comment by kwonye** [comment]
Crash. I changed a discussion setting, backed out of settings, and was navigating around the app.

```
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime: FATAL EXCEPTION: main
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime: Process: org.wordpress.android, PID: 32417
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime: java.lang.IllegalStateException: Fragment SiteSettingsFragment{ad94eed} not attached to Activity
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at android.app.Fragment.getResources(Fragment.java:808)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at android.app.Fragment.getString(Fragment.java:830)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at org.wordpress.android.ui.prefs.SiteSettingsFragment.getPrivacySummary(SiteSettingsFragment.java:669)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at org.wordpress.android.ui.prefs.SiteSettingsFragment.setPreferencesFromSiteSettings(SiteSettingsFragment.java:518)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at org.wordpress.android.ui.prefs.SiteSettingsFragment.onSettingsUpdated(SiteSettingsFragment.java:389)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at org.wordpress.android.ui.prefs.SiteSettingsInterface$1.run(SiteSettingsInterface.java:501)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at android.app.Activity.runOnUiThread(Activity.java:5511)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at org.wordpress.android.ui.prefs.SiteSettingsInterface.notifyUpdatedOnUiThread(SiteSettingsInterface.java:498)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at org.wordpress.android.ui.prefs.DotComSiteSettings$3.onResponse(DotComSiteSettings.java:118)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at org.wordpress.android.ui.prefs.DotComSiteSettings$3.onResponse(DotComSiteSettings.java:107)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at com.wordpress.rest.RestRequest.deliverResponse(RestRequest.java:74)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at com.wordpress.rest.RestRequest.deliverResponse(RestRequest.java:18)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:99)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at android.os.Handler.handleCallback(Handler.java:739)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at android.os.Handler.dispatchMessage(Handler.java:95)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at android.os.Looper.loop(Looper.java:148)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at android.app.ActivityThread.main(ActivityThread.java:5417)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at java.lang.reflect.Method.invoke(Native Method)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
11-18 16:13:02.055 32417-32417/org.wordpress.android E/AndroidRuntime:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

**Comment by kwonye** [comment]
Unchecking "Allow Comments" still allows comments on dubkwon.wordpress.com

**Comment by tonyr59h** [comment]
> I noticed that there is no consistency to capitalization.

I'm going to have @mattmiklic take a look through the design once this is merged. There will be a cleanup PR to address issues like this.

**Comment by tonyr59h** [comment]
> Crash.

Addressed in a65770f478a9b0df31b0c47a10dfa8d4c254f4a0

**Comment by kwonye** [comment]
> Unchecking "Allow Comments" still allows comments on dubkwon.wordpress.com

Turns out I was unclear about the setting. "Allow comments" only applies to new posts. @tonyr59h has made a note to address the confusion in the polish PR, so effectively addressed.

**Comment by kwonye** [comment]
Another crash (on self-hosted site):
```
java.lang.RuntimeException: Unable to destroy activity {org.wordpress.android/org.wordpress.android.ui.prefs.BlogPreferencesActivity}: 
java.lang.NullPointerException: Attempt to invoke interface method 'java.util.Iterator java.util.List.iterator()' on a null object reference
                                                                         at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3831)
                                                                         at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:3849)
                                                                         at android.app.ActivityThread.-wrap5(ActivityThread.java)
                                                                         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1398)
                                                                         at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                         at android.os.Looper.loop(Looper.java:148)
                                                                         at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                         at java.lang.reflect.Method.invoke(Native Method)
                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                      Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'java.util.Iterator java.util.List.iterator()' on a null object reference
                                                                         at org.wordpress.android.models.SiteSettingsModel.serializeToDatabase(SiteSettingsModel.java:296)
                                                                         at org.wordpress.android.datasets.SiteSettingsTable.saveSettings(SiteSettingsTable.java:86)
                                                                         at org.wordpress.android.ui.prefs.SiteSettingsInterface.saveSettings(SiteSettingsInterface.java:112)
                                                                         at org.wordpress.android.ui.prefs.SelfHostedSiteSettings.saveSettings(SelfHostedSiteSettings.java:79)
                                                                         at org.wordpress.android.ui.prefs.SiteSettingsFragment.onDestroy(SiteSettingsFragment.java:149)
                                                                         at android.app.Fragment.performDestroy(Fragment.java:2436)
                                                                         at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1109)
                                                                         at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1159)
                                                                         at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1141)
                                                                         at android.app.FragmentManagerImpl.dispatchDestroy(FragmentManager.java:1992)
                                                                         at android.app.FragmentController.dispatchDestroy(FragmentController.java:218)
                                                                         at android.app.Activity.performDestroy(Activity.java:6406)
                                                                         at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1142)
                                                                         at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3818)
                                                                         at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:3849) 
                                                                         at android.app.ActivityThread.-wrap5(ActivityThread.java) 
                                                                         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1398) 
                                                                         at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                         at android.os.Looper.loop(Looper.java:148) 
                                                                         at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                         at java.lang.reflect.Method.invoke(Native Method) 
                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
```

**Comment by kwonye** [comment]
On self-hosted, setting `Send Pingbacks` to either on or off turns off Default article settings > Attempt to notify any blogs linked to from the article on WP Dashboard.

**Comment by mattmiklic** [comment]
> I noticed that there is no consistency to capitalization. There's "Allow Comments" and below there is "Links in comments".

I spoke with @drw158 about this while designing this section and he noted that the Material Design guidelines suggest sentence case for all labels:

> Capitalize only the first word of each label, and proper nouns.
— https://www.google.com/design/spec/patterns/settings.html#settings-grouping-settings

So let's standardize on this.

**Comment by tonyr59h** [comment]
> Another crash (on self-hosted site)

Addressed in e92f49158b6acd2324def806b235ef380bc5b87d

> On self-hosted, setting Send Pingbacks to either on or off turns off Default article settings

Addressed in bae5cb63467a38be50a3a3d8195a1763898c4f75

**Comment by kwonye** [comment]
Added #3434 as it is required for settings to go into `develop`

**Comment by kwonye** [comment]
Crash when accessing settings while in airplane mode: 

```
FATAL EXCEPTION: main  Process: org.wordpress.android, PID: 7896
                                                                     java.lang.RuntimeException: Unable to destroy activity {org.wordpress.android/org.wordpress.android.ui.prefs.BlogPreferencesActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void org.wordpress.android.ui.prefs.SiteSettingsInterface.saveSettings()' on a null object reference
                                                                         at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3831)
                                                                         at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:3849)
                                                                         at android.app.ActivityThread.-wrap5(ActivityThread.java)
                                                                         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1398)
                                                                         at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                         at android.os.Looper.loop(Looper.java:148)
                                                                         at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                         at java.lang.reflect.Method.invoke(Native Method)
                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                      Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void org.wordpress.android.ui.prefs.SiteSettingsInterface.saveSettings()' on a null object reference
                                                                         at org.wordpress.android.ui.prefs.SiteSettingsFragment.onDestroy(SiteSettingsFragment.java:149)
                                                                         at android.app.Fragment.performDestroy(Fragment.java:2436)
                                                                         at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1109)
                                                                         at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1159)
                                                                         at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1141)
                                                                         at android.app.FragmentManagerImpl.dispatchDestroy(FragmentManager.java:1992)
                                                                         at android.app.FragmentController.dispatchDestroy(FragmentController.java:218)
                                                                         at android.app.Activity.performDestroy(Activity.java:6406)
                                                                         at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1142)
                                                                         at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3818)
                                                                         at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:3849) 
                                                                         at android.app.ActivityThread.-wrap5(ActivityThread.java) 
                                                                         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1398) 
                                                                         at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                         at android.os.Looper.loop(Looper.java:148) 
                                                                         at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                         at java.lang.reflect.Method.invoke(Native Method) 
                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
```

**Comment by kwonye** [comment]
Great job! Not able to break it anymore :stuck_out_tongue: 

:shipit:

**Pull request merged**